modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/ExpressionUtilsImpl.java

```diff
@@ -93,56 +93,138 @@ public Object evaluate(final Node question, final Map<String, Object> values, fi
+            this.endMarker = this.isArrayArgument ? END_MARKER_ARRAY : END_MARKER_SINGLE;
+            this.start = this.expression.indexOf(this.startMarker);
+            this.end = this.expression.indexOf(this.endMarker);
+        }
+
+        public ParsedExpression parse()
+        {
+            while (hasNextArgument()) {
+                parseNextArgument();
             }
+            return new ParsedExpression(this.questions, this.expression, this.missingValue);
+
+        }
+
+        private boolean hasNextArgument()
+        {
+            return this.start > -1 && this.end > -1;
+        }
+
+        private boolean isOptionalQuestion(boolean hasDefault, int defaultStart)
+        {
+            int endIndex = hasDefault ? defaultStart : this.end;
+            String potentialOptionalString = this.expression.substring(endIndex - OPTIONAL_MARKER.length(), endIndex);
+
+            return OPTIONAL_MARKER.equals(potentialOptionalString);
+        }
+
+        private String getQuestionName(boolean hasDefault, int defaultStart, boolean isOptional)
+        {
+            String questionName = this.expression.substring(this.start + this.startMarker.length(),
+                hasDefault ? defaultStart : this.end);
+            if (isOptional) {
+                questionName = questionName.substring(0, questionName.length() - OPTIONAL_MARKER.length());
+            }
+            return questionName;
+        }
+
+        private String getDefaultValue(boolean hasDefault, int defaultStart)
+        {
+            return hasDefault ? this.expression.substring(defaultStart + DEFAULT_MARKER.length(), this.end) : null;
+        }
+
+        private void parseNextArgument()
+        {
+            // Parse the question name and default value if present.
+            int defaultStart = this.expression.indexOf(DEFAULT_MARKER, this.start);
+            boolean hasDefault = defaultStart > -1 && defaultStart < this.end;
+
+            boolean isOptional = isOptionalQuestion(hasDefault, defaultStart);
+            String questionName = getQuestionName(hasDefault, defaultStart, isOptional);
 
             // Insert this question into the list of arguments
-            if (!questions.containsKey(questionName)) {
-                ExpressionArgument arg = new ExpressionArgument("arg" + questions.size(),
-                    getQuestionValue(questionName, values, defaultValue));
-                if (arg.getValue() == null) {
-                    missingValue = true;
+            if (!this.questions.containsKey(questionName)) {
+                Object questionValue = getQuestionValue(questionName, this.questionValues,
+                    getDefaultValue(hasDefault, defaultStart), this.isArrayArgument);
+
+                if (questionValue == null) {
+                    if (!isOptional) {
+                        this.missingValue = true;
+                    } else if (isOptional && this.isArrayArgument) {
```

Comment:
We already know it is optional:
```suggestion
                    } else if (this.isArrayArgument) {
```